spec/imsc-hrm.html

```diff
@@ -294,11 +292,16 @@ <h2>General</h2>
       <p>The Presentation Compositor SHALL start rendering E<sub>n</sub>:</p>
 
       <ul>
-        <li>at the presentation time of E<sub>0</sub> minus Initial Painting Delay (IPD), if n = 0; or</li>
-
-        <li>at the presentation time of E<sub>n-1</sub>, if n &gt; 0.</li>
+        <li>at the presentation time of E<sub>m</sub>, where E<sub>m</sub> is the closest ISD for which m&lt;n and E<sub>m-1</sub>
```

Comment:
Shouldn't it be E<sub>m+1</sub> is not empty? Otherwise if the sequence is:

E<sub>n-x-1</sub>: Empty
E<sub>n-x</sub>: Has visible content
E<sub>n-x+1</sub>: Empty
E<sub>...</sub>: [maybe more empty ones]
E<sub>n</sub>: Has visible content

We want m == n-x here, I think. But E<sub>n-x-1</sub> is empty, so the current text excludes m == n-x which is an unexpected result.

Reply:
Urgh. Thanks for catching this. Will fix.

Reply:
Actually we want `m == n-x+1`: assuming there is no sequences of empty ISDs, the rendering of `E_n` should start at the presentation time of `E_(n-1)`, even if `E_(n-1)` is empty.

Reply:
I thought that's what the current text says, that causes the problem when there's a 2 frame (say) empty ISD, and that's what we wanted to resolve. That's why I thought the conclusion was that we want the "render timer" to begin at `n-x` rather than `n-x+1`.

Reply:
Given the following sequence:

```
E5: Empty
E6: Has visible content
E7: Empty
E8: Empty
E9: Empty
E10: Has visible content
```

We want the rendering of E10 to start at the presentation time of E7 and not E9.

Reply:
This is where we disagree, perhaps: I think we need it to be E6 not E7.

Reply:
As currently specified, the rendering of document E_n starts at the presentation time of document E_(n-1), even if E_(n-1) is empty. This is necessary since there is a cost to E_(n-1), even if E_(n-1) is empty: the root container has to be cleared. Do you agree?

If so, all that we are doing is making sure that multiple empty ISDs in sequence are considered as one.

Reply:
Yes, right, but I don't think that solves the problem that's been presented, which is that a single 2 frame empty ISD causes the following ISD to fail the HRM.

I think we need to imagine that "clearing all the pixels" is an activity that can be done independently of "preparing presentation of a non-empty ISD", and that it has a known fixed cost.

Reply:
>Yes, right, but I don't think that solves the problem that's been presented, 

The problem was two-fold based on the files I have seen:

* E10 was forced to clear the root container even though it had been cleared by E7
* E10 only had the duration of E9 to render, even though rendering could have started at E7 -- this is particularly acute if E9 is very short.

This PR is intended to address the two issues above. Makes sense?

Reply:
Yes, makes sense that it would resolve those issues, but what doesn't make sense is how we get to a situation where there are multiple consecutive empty ISDs, given that the `[resolve timing]` procedure is only supposed to generate time coordinates where some element becomes active or inactive. If two consecutive ISDs are both empty, doesn't that imply that no element became active between the first and second? And since the ISDs are empty, there wasn't anything to become inactive that was already active.

Is it not correct that we need a solution to short (single) empty ISDs causing files to fail HRM validation?

Reply:
> , but what doesn't make sense is how we get to a situation where there are multiple consecutive empty ISDs,

Two scenarios come to mind:

* the `[resolve timing]` does not preclude generating multiple empty ISDs in sequence. For example, `set` elements can generate time coordinates without causing visual differences and without being present in the ISD (by definition):

```
<region xml:id="r1" begin="0s" end="1s" tts:background="transparent">
   <set tts:background="transparent" begin="0.5s" end="1s">
</region>
```

* there is a temporal gap between documents in a sequence of documents, resulting effectively in an empty ISD.

There might be more.

Reply:
@nigelmegitt  added illustration